File: src/main/java/xyz/oribuin/lilori/commands/moderation/CmdBan.java
package xyz.oribuin.lilori.commands.moderation;

import net.dv8tion.jda.api.entities.Member;
import xyz.oribuin.lilori.utilities.command.Command;
import xyz.oribuin.lilori.utilities.command.CommandEvent;
import xyz.oribuin.lilori.persist.Settings;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.Permission;
import xyz.oribuin.lilori.utilities.command.Command;
import xyz.oribuin.lilori.utilities.command.CommandEvent;

import java.awt.*;
import java.sql.Time;
import java.util.ArrayList;
import java.util.concurrent.TimeUnit;

public class CmdBan extends Command {
    public CmdBan() {
        this.name = "Ban";
        this.aliases = new String[]{"Expel"};
        this.help = "Ban a member from entering your guild";
        this.cooldown = 2;
        this.category = new Command.Category("Moderation");
        this.guildOnly = true;
        this.arguments = "[@User] [Messages] [Reason]";

        this.userPermissions = new Permission[]{Permission.MESSAGE_MANAGE, Permission.BAN_MEMBERS};
        this.botPermissions = new Permission[]{Permission.MESSAGE_MANAGE, Permission.BAN_MEMBERS};
    }

    @Override
    protected void execute(CommandEvent event) {
        String[] args = event.getMessage().getContentRaw().split(" ");

        if (args.length <= 3) {
            event.deleteCmd(10, TimeUnit.SECONDS);
            event.timedReply(event.getAuthor().getAsMention() + ", Correct Format: ;ban [@User] [Messages] [Reason]", 10, TimeUnit.SECONDS);
            return;
        }

        if (event.getMessage().getMentionedMembers().get(0) == null) {
            event.deleteCmd(10, TimeUnit.SECONDS);
            event.timedReply(event.getAuthor().getAsMention() + ", Please include a user to ban.", 10, TimeUnit.SECONDS);
            return;
        }

        Member mentioned = event.getMessage().getMentionedMembers().get(0);

        if (mentioned.isOwner() || event.isHigher(mentioned, event.getGuild().getMember(event.getAuthor())) || event.isBot(mentioned)) {
            event.deleteCmd(10, TimeUnit.SECONDS);
            event.timedReply(event.getAuthor().getAsMention() + ", You cannot ban this user due to rank hierarchy.", 10, TimeUnit.SECONDS);
            return;
        }

        int messages;
        String reason = event.getMessage().getContentRaw().substring(args[0].length() + 1 + args[1].length() + 1 + args[2].length() + 1);

        try {
            messages = Integer.parseInt(args[2]);
        } catch (NumberFormatException ex) {
            event.deleteCmd(10, TimeUnit.SECONDS);
            event.timedReply(event.getAuthor().getAsMention() + ", Please include a correct number of messages.", 10, TimeUnit.SECONDS);
            return;
        }

        event.banMember(mentioned.getUser(), messages, reason);

        event.deleteCmd();
        event.reply(event.getAuthor().getAsMention() + ", You have banned `" + mentioned.getUser().getAsTag() + "`");
    }
}

File: src/main/kotlin/xyz/oribuin/lilori/commands/global/moderation/CmdBan.kt
package xyz.oribuin.lilori.commands.global.moderation

import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.exceptions.HierarchyException
import xyz.oribuin.lilori.Settings
import xyz.oribuin.lilori.handler.Command
import xyz.oribuin.lilori.handler.CommandEvent

class CmdBan : Command() {
    init {
        name = "Ban"
        aliases = listOf("bean")
        description = "Ban a member from a server!"
        userPermissions = arrayOf(Permission.BAN_MEMBERS)
        botPermissions = arrayOf(Permission.BAN_MEMBERS)
    }

    override fun executeCommand(event: CommandEvent) {
        // Define all the variables
        event.deleteCmd()

        val args = event.message.contentRaw.split(" ").toTypedArray()
        val msgTime = event.message.timeCreated

        if (args.size < 4) {
            sendInvalidArgs(event)
            return
        }

        if (event.message.mentionedMembers.size == 0) {
            sendInvalidUsersMentioned(event)
            return
        }

        val member = event.message.mentionedMembers[0]

        if (member == null) {
            sendInvalidUsersMentioned(event)
            return
        }

        if (isHigher(event)) {
            sendRankHierarchy(event)
            return
        }

        try {
            val delDays = args[2].toInt()

            if (delDays < 0 || delDays > 100) {
                sendInvalidNumber(event)
                return
            }

            val reason = event.message.contentRaw.substring(args[0].length + args[1].length + args[2].length + 3)

            member.ban(delDays, reason).queue()

            val embedBuilder = EmbedBuilder()
                    .setAuthor("\uD83D\uDC94 Banned User " + member.user.asTag)
                    .setDescription("""Successfully banned ${member.user.asTag} (${member.id}) at ${msgTime.hour}h, ${msgTime.minute}m & ${msgTime.second}s on
                            ${msgTime.dayOfMonth} ${msgTime.month.name.toLowerCase()} for $reason""".trimMargin())
                    .setColor(Settings.EMBED_COLOR)
                    .setFooter("get bonked", "https://img.oribuin.xyz/bot-images/bonk.gif")

            event.channel.sendMessage(event.author.asMention).embed(embedBuilder.build()).queue()

        } catch (ex: NumberFormatException) {
            sendInvalidNumber(event)
        } catch (ex: HierarchyException) {
            sendRankHierarchy(event)
        }
    }

    private fun sendInvalidArgs(event: CommandEvent) {
        val embed = EmbedBuilder()
                .setAuthor("Invalid Arguments")
                .setDescription("You have provided invalid arguments for the command!")
                .setColor(Settings.EMBED_COLOR)
                .setAuthor("~ban @<User> <Delete-Days> <Reason>")
        event.channel.sendMessage(event.author.asMention).embed(embed.build()).queue()
    }

    private fun sendInvalidUsersMentioned(event: CommandEvent) {
        val embed = EmbedBuilder()
                .setAuthor("\uD83D\uDC94 Invalid User")
                .setDescription("Please include a valid user!")
                .setColor(Settings.EMBED_COLOR)
                .setAuthor("~ban @<User> <Delete-Days> <Reason>")
        event.channel.sendMessage(event.author.asMention).embed(embed.build()).queue()
    }

    private fun sendInvalidNumber(event: CommandEvent) {
        val embed = EmbedBuilder()
                .setAuthor("\uD83D\uDC94 Invalid Number")
                .setDescription("Please make your third argument a number between 0-100!")
                .setColor(Settings.EMBED_COLOR)
                .setAuthor("~ban @<User> <Delete-Days> <Reason>")
        event.channel.sendMessage(event.author.asMention).embed(embed.build()).queue()
    }

    private fun sendRankHierarchy(event: CommandEvent) {
        val embed = EmbedBuilder()
                .setAuthor("\uD83D\uDC94 Can't ban that user")
                .setDescription("Cannot ban this user due to hierarchy!")
                .setColor(Settings.EMBED_COLOR)
                .setAuthor("~ban @<User> <Delete-Days> <Reason>")
        event.channel.sendMessage(event.author.asMention).embed(embed.build()).queue()
    }

    private fun isHigher(event: CommandEvent): Boolean {
        val mentioned = event.message.mentionedMembers[0]
        return if (mentioned === event.member) true else if (mentioned.isOwner) true else if (mentioned.isFake) true else if (mentioned.roles[0].position >= event.selfMember.roles[0].position) true else if (mentioned.roles[0].position >= event.member!!.roles[0].position) true else mentioned.user.isBot
    }
}